File: tests/test_reconciler.py
from __future__ import annotations

import pytest

from aes67_nmos_bridge.models import (
    BridgeConfig,
    DaemonState,
    DaemonStream,
    ReceiverConfig,
    SenderConfig,
)
from aes67_nmos_bridge.ownership import managed_name
from aes67_nmos_bridge.reconciler import ReconcileError, Reconciler


class FakeDaemon:
    def __init__(self, state: DaemonState):
        self.state = state
        self.calls: list[tuple[str, int, dict | None]] = []

    def get_state(self) -> DaemonState:
        return self.state

    def put_sender(self, daemon_id: int, payload: dict) -> None:
        self.calls.append(("put_sender", daemon_id, payload))

    def delete_sender(self, daemon_id: int) -> None:
        self.calls.append(("delete_sender", daemon_id, None))

    def put_receiver(self, daemon_id: int, payload: dict) -> None:
        self.calls.append(("put_receiver", daemon_id, payload))

    def delete_receiver(self, daemon_id: int) -> None:
        self.calls.append(("delete_receiver", daemon_id, None))


def test_creates_missing_configured_streams() -> None:
    config = BridgeConfig(
        senders=(SenderConfig("main", 1, "Main", (0, 1)),),
        receivers=(ReceiverConfig("return", 2, "Return", (2, 3), sdp=_sdp()),),
    )
    daemon = FakeDaemon(DaemonState())

    report = Reconciler(config, daemon).reconcile()

    assert [operation.action for operation in report.operations] == ["create", "create"]
    assert daemon.calls[0][0:2] == ("put_sender", 1)
    assert daemon.calls[0][2]["name"] == "NMOS[default]/sender/main Main"
    assert daemon.calls[1][0:2] == ("put_receiver", 2)
    assert daemon.calls[1][2]["use_sdp"] is True


def test_updates_drifted_managed_sender() -> None:
    config = BridgeConfig(senders=(SenderConfig("main", 1, "Main", (0, 1), codec="L24"),))
    existing = DaemonStream(
        "sender",
        1,
        {
            "id": 1,
            "name": managed_name("default", "sender", "main", "Main"),
            "enabled": True,
            "io": "Audio Device",
            "codec": "L16",
            "address": "",
            "max_samples_per_packet": 48,
            "ttl": 15,
            "payload_type": 98,
            "dscp": 34,
            "refclk_ptp_traceable": True,
            "map": [0, 1],
        },
    )
    daemon = FakeDaemon(DaemonState(senders=(existing,)))

    report = Reconciler(config, daemon).reconcile()

    assert [(op.action, op.reason) for op in report.operations] == [
        ("update", "daemon stream drifted")
    ]
    assert daemon.calls == [("put_sender", 1, report.operations[0].payload)]
    assert daemon.calls[0][2]["codec"] == "L24"


def test_deletes_orphaned_managed_streams_but_keeps_unmanaged() -> None:
    orphan = DaemonStream(
        "sender",
        7,
        {"id": 7, "name": managed_name("default", "sender", "old", "Old")},
    )
    unmanaged = DaemonStream("sender", 8, {"id": 8, "name": "Manual source"})
    daemon = FakeDaemon(DaemonState(senders=(orphan, unmanaged)))

    report = Reconciler(BridgeConfig(), daemon).reconcile()

    assert [(op.action, op.side, op.daemon_id) for op in report.operations] == [
        ("delete", "sender", 7)
    ]
    assert daemon.calls == [("delete_sender", 7, None)]


def test_refuses_to_overwrite_unmanaged_stream_in_configured_slot() -> None:
    config = BridgeConfig(senders=(SenderConfig("main", 1, "Main", (0, 1)),))
    unmanaged = DaemonStream("sender", 1, {"id": 1, "name": "Manual source"})
    daemon = FakeDaemon(DaemonState(senders=(unmanaged,)))

    with pytest.raises(ReconcileError, match="occupied by unmanaged stream"):
        Reconciler(config, daemon).plan()


def test_disabled_receiver_deletes_owned_sink() -> None:
    config = BridgeConfig(
        receivers=(ReceiverConfig("ret", 2, "Return", (0, 1), sdp=_sdp(), enabled=False),),
    )
    existing = DaemonStream(
        "receiver",
        2,
        {"id": 2, "name": managed_name("default", "receiver", "ret", "Return")},
    )
    daemon = FakeDaemon(DaemonState(receivers=(existing,)))

    report = Reconciler(config, daemon).reconcile()

    assert [(op.action, op.side, op.daemon_id) for op in report.operations] == [
        ("delete", "receiver", 2)
    ]
    assert daemon.calls == [("delete_receiver", 2, None)]


def test_disabled_receiver_does_not_create_sink() -> None:
    config = BridgeConfig(
        receivers=(ReceiverConfig("ret", 2, "Return", (0, 1), sdp=_sdp(), enabled=False),),
    )
    daemon = FakeDaemon(DaemonState())

    report = Reconciler(config, daemon).reconcile()

    assert report.operations == ()
    assert daemon.calls == []


def test_disabled_receiver_leaves_unmanaged_slot_untouched() -> None:
    config = BridgeConfig(
        receivers=(ReceiverConfig("ret", 2, "Return", (0, 1), sdp=_sdp(), enabled=False),),
    )
    unmanaged = DaemonStream("receiver", 2, {"id": 2, "name": "Manual sink"})
    daemon = FakeDaemon(DaemonState(receivers=(unmanaged,)))

    report = Reconciler(config, daemon).reconcile()

    assert report.operations == ()
    assert daemon.calls == []


def test_dry_run_does_not_call_daemon_mutations() -> None:
    config = BridgeConfig(senders=(SenderConfig("main", 1, "Main", (0, 1)),))
    daemon = FakeDaemon(DaemonState())

    report = Reconciler(config, daemon).reconcile(dry_run=True)

    assert report.changed is True
    assert report.dry_run is True
    assert daemon.calls == []


def _sdp() -> str:
    return (
        "v=0\n"
        "o=- 1 0 IN IP4 192.0.2.10\n"
        "s=Remote AES67 Source\n"
        "c=IN IP4 239.2.0.10/15\n"
        "t=0 0\n"
        "m=audio 5004 RTP/AVP 98\n"
        "a=rtpmap:98 L24/48000/2\n"
    )

File: tests/test_reconciler.cpp
#include <catch2/catch_test_macros.hpp>

#include <tuple>
#include <vector>

#include "bridge/ownership.h"
#include "bridge/reconciler.h"

using namespace bridge;

namespace {

web::json::value named(const std::string& name) {
    web::json::value v;
    v[U("name")] = web::json::value::string(utility::conversions::to_string_t(name));
    return v;
}

web::json::value sender_payload(const std::string& name, const std::string& codec) {
    auto v = named(name);
    v[U("codec")] = web::json::value::string(utility::conversions::to_string_t(codec));
    return v;
}

DaemonStream stream(Side side, int id, const std::string& name) {
    DaemonStream s;
    s.side = side;
    s.id = id;
    s.name = name;
    s.payload = named(name);
    return s;
}

DesiredStream desired_sender(int daemon_id, const std::string& nmos_id, bool enabled,
                             const web::json::value& payload = {}) {
    DesiredStream d;
    d.side = Side::sender;
    d.daemon_id = daemon_id;
    d.nmos_id = nmos_id;
    d.enabled = enabled;
    d.name = managed_name("default", Side::sender, nmos_id, "Label");
    d.payload = payload;
    return d;
}

DesiredStream desired_receiver(int daemon_id, const std::string& nmos_id, bool enabled,
                               const web::json::value& payload = {}) {
    DesiredStream d;
    d.side = Side::receiver;
    d.daemon_id = daemon_id;
    d.nmos_id = nmos_id;
    d.enabled = enabled;
    d.name = managed_name("default", Side::receiver, nmos_id, "Label");
    d.payload = payload;
    return d;
}

class FakeDaemon : public IDaemon {
public:
    DaemonState state;
    std::vector<std::tuple<std::string, int>> calls;

    DaemonState get_state() override { return state; }
    void put_source(int id, const web::json::value&) override { calls.emplace_back("put_source", id); }
    void delete_source(int id) override { calls.emplace_back("delete_source", id); }
    void put_sink(int id, const web::json::value&) override { calls.emplace_back("put_sink", id); }
    void delete_sink(int id) override { calls.emplace_back("delete_sink", id); }
};

}  // namespace

TEST_CASE("creates missing configured streams") {
    FakeDaemon daemon;
    Reconciler reconciler("default", daemon);

    std::vector<DesiredStream> desired = {
        desired_sender(1, "main", true, sender_payload(managed_name("default", Side::sender, "main", "Label"), "L24")),
        desired_receiver(2, "ret", true, named(managed_name("default", Side::receiver, "ret", "Label"))),
    };

    const auto report = reconciler.reconcile(desired);
    REQUIRE(report.ops.size() == 2);
    CHECK(report.errors.empty());
    CHECK(daemon.calls == std::vector<std::tuple<std::string, int>>{{"put_source", 1}, {"put_sink", 2}});
}

TEST_CASE("updates a drifted managed sender") {
    FakeDaemon daemon;
    daemon.state.sources[1] =
        stream(Side::sender, 1, managed_name("default", Side::sender, "main", "Label"));
    daemon.state.sources[1].payload =
        sender_payload(managed_name("default", Side::sender, "main", "Label"), "L16");
    Reconciler reconciler("default", daemon);

    std::vector<DesiredStream> desired = {desired_sender(
        1, "main", true, sender_payload(managed_name("default", Side::sender, "main", "Label"), "L24"))};

    const auto report = reconciler.reconcile(desired);
    REQUIRE(report.ops.size() == 1);
    CHECK(report.ops[0].action == PlannedOp::Action::update);
    CHECK(daemon.calls == std::vector<std::tuple<std::string, int>>{{"put_source", 1}});
}

TEST_CASE("no drift means no operations") {
    FakeDaemon daemon;
    const auto payload = sender_payload(managed_name("default", Side::sender, "main", "Label"), "L24");
    daemon.state.sources[1] =
        stream(Side::sender, 1, managed_name("default", Side::sender, "main", "Label"));
    daemon.state.sources[1].payload = payload;
    Reconciler reconciler("default", daemon);

    const auto report = reconciler.reconcile({desired_sender(1, "main", true, payload)});
    CHECK(report.ops.empty());
    CHECK(daemon.calls.empty());
}

TEST_CASE("deletes orphaned managed stream but keeps unmanaged") {
    FakeDaemon daemon;
    daemon.state.sources[7] =
        stream(Side::sender, 7, managed_name("default", Side::sender, "old", "Old"));
    daemon.state.sources[8] = stream(Side::sender, 8, "Manual source");
    Reconciler reconciler("default", daemon);

    const auto report = reconciler.reconcile({});
    REQUIRE(report.ops.size() == 1);
    CHECK(report.ops[0].action == PlannedOp::Action::remove);
    CHECK(report.ops[0].daemon_id == 7);
    CHECK(daemon.calls == std::vector<std::tuple<std::string, int>>{{"delete_source", 7}});
}

TEST_CASE("refuses to overwrite unmanaged stream in a configured slot") {
    FakeDaemon daemon;
    daemon.state.sources[1] = stream(Side::sender, 1, "Manual source");
    Reconciler reconciler("default", daemon);

    const auto report = reconciler.reconcile({desired_sender(1, "main", true, named("x"))});
    CHECK(report.ops.empty());
    REQUIRE(report.errors.size() == 1);
    CHECK(daemon.calls.empty());
}

TEST_CASE("refuses to overwrite a stream owned by another namespace") {
    FakeDaemon daemon;
    daemon.state.sources[1] =
        stream(Side::sender, 1, managed_name("truck-b", Side::sender, "main", "Label"));
    Reconciler reconciler("default", daemon);

    const auto report = reconciler.reconcile({desired_sender(1, "main", true, named("x"))});
    CHECK(report.ops.empty());
    REQUIRE(report.errors.size() == 1);
}

TEST_CASE("disabled receiver deletes the owned sink") {
    FakeDaemon daemon;
    daemon.state.sinks[2] =
        stream(Side::receiver, 2, managed_name("default", Side::receiver, "ret", "Label"));
    Reconciler reconciler("default", daemon);

    const auto report = reconciler.reconcile({desired_receiver(2, "ret", false)});
    REQUIRE(report.ops.size() == 1);
    CHECK(report.ops[0].action == PlannedOp::Action::remove);
    CHECK(daemon.calls == std::vector<std::tuple<std::string, int>>{{"delete_sink", 2}});
}

TEST_CASE("disabled receiver leaves an unmanaged sink untouched") {
    FakeDaemon daemon;
    daemon.state.sinks[2] = stream(Side::receiver, 2, "Manual sink");
    Reconciler reconciler("default", daemon);

    const auto report = reconciler.reconcile({desired_receiver(2, "ret", false)});
    CHECK(report.ops.empty());
    CHECK(report.errors.empty());
    CHECK(daemon.calls.empty());
}
